Add Zstd support as optional and compile-time low ratio report

kkent030315 opened the pull request:
Zstd (*ZStandard*) is a fast real-time compression algorithm.

~~I've added Zstd support in compression/decompression as optional. Users can select which one to use by specifying feature flags: `zstd` and `deflate`.~~

```toml
# Cargo.toml
[dependencies]
include-flate = { version = "...", default-features = false, features = ["zstd"] }
# or
include-flate = { version = "...", default-features = false, features = ["deflate"] } # this is same as default.
```

In summary, while DEFLATE is a venerable and widely adopted algorithm, Zstd provides significant advantages in speed and compression ratio, would make it a strong choice for crates or binaries that can support it.

### ~~Testing this PR~~

~~Since I've modified `Cargo.toml` to not use unnecessary dependency, multiple vectors of tests needed:~~

```
cargo test --all
```

```
cargo test --no-default-features --features deflate
```

```
cargo test --no-default-features --features zstd
```

SOF3 commented:
in that case, consider using the CI pipeline for different features too?

SOF3 commented on src/lib.rs:
```diff
@@ -114,10 +120,22 @@ macro_rules! flate {
-    dec.read_to_end(&mut ret)
-        .expect("Compiled DEFLATE buffer was corrupted");
+
+    #[cfg(feature = "zstd")]
```
Cargo features are additive. What if someone uses two libraries, each requesting a different feature?
I think the compression algorithm should be configured at call site instead of compilation flags. Features should only affect whether an algorithm is supported, not whether it is used

kkent030315 replied:
Thank you for your feedback. Firstly, I agree with your opinion overall. This is good point, I were seeking to how this feature should be implemented. I choose the Cargo features because I believe there should not be the unnecessary crate in the dependency. This is more important if it becomes an C-binding crate like [`zstd-rs`](https://github.com/gyscos/zstd-rs). The approach is similar to the [`zip-rs`](https://github.com/zip-rs/zip/blob/3e88fe66c941d411cff5cf49778ba08c2ed93801/Cargo.toml#L36-L42).

Alternatively, we can implement a wrapper for `Decoder` and `Encoder`. I am not sure that way would satisfy you. Could you tell the specifics of what you're think ideal?

kkent030315 replied:
Instead of Cargo features, we can do something like this to make the compression method selective for each files, however, always depending on the potentially-unwanted crates as a result, if acceptable:
```rust
flate!(pub static DATA: [u8] from "assets/009f.dat"); // default, DEFLATE
flate!(pub static DATA: [u8] from "assets/009f.dat" include_flate::Compression::Zstd); // Use Zstd for this file spcifically
flate!(pub static DATA: [u8] from "assets/009f.dat" include_flate::Compression::Deflate); // Explicitly use DEFLATE.
```

SOF3 replied:
Looks good to me, but perhaps we could omit the namespace as well since it's just a keyword read by the proc-macro.

kkent030315 replied:
Sure thing, what about a some string?
```rust
flate!(pub static DATA: [u8] from "assets/009f.dat"); // default, DEFLATE
flate!(pub static DATA: [u8] from "assets/009f.dat" "zstd"); // Use Zstd for this file spcifically
flate!(pub static DATA: [u8] from "assets/009f.dat" "deflate"); // Explicitly use DEFLATE.
```

SOF3 replied:
or maybe just a keyword?

```rust
flate!(pub static DATA: [u8] from "assets/009f.dat"); // default, DEFLATE
flate!(pub static DATA: [u8] from "assets/009f.dat" with zstd); // Use Zstd for this file spcifically
flate!(pub static DATA: [u8] from "assets/009f.dat" with deflate); // Explicitly use DEFLATE.
```

```rust
($vis:vis static $ident:ident: $ty:ty from $file:literal $(with $algo:ident)?
```

kkent030315 commented:
Well, It's done and awaiting for your @SOF3 review.

### Changelog

- [x] **Added**: Zstd support for compression.
  - I've added custom encoders, `FlateEncoder` and `FlateDecoder`, for each of the current and future compression methods that the encoders/decoders need to wrap. That means, it is just to write an wrapper for both `FlateEncoder` and `FlateDecoder` and adding some values into some enums are enough to add custom compression method in further. And of course, I've improved tests, so additional compression methods could be tested more precisely.
- [x] **Added**: `include-flate-compress` in `compress` folder in the root directory of this repository, for commonality.
- [x] **Added**: `FlateArgs` for extendable/descriptive `TokenStream` input.
- [x] **Improved**: Entire tests are improved for additional checks regarding compression.
- [x] **Improved**: `test_util.rs` has been reworked for reliable code.
- [x] **Improved**: Better error message in decompression failure in `include_flate::decode` function.

The compression features are selectable for each `flate!` macros, or `DEFLATE` by default:

```rust
flate!(pub static DATA: [u8] from "assets/009f.dat"); // default, DEFLATE
flate!(pub static DATA: [u8] from "assets/009f.dat" with zstd); // Use Zstd for this file spcifically
flate!(pub static DATA: [u8] from "assets/009f.dat" with deflate); // Explicitly use DEFLATE.
```

kkent030315 commented:
> in that case, consider using the CI pipeline for different features too?

As of now, I've pivoted to not use Cargo feature, CI changes shouldn't required. (Reverted in [957c2bf](https://github.com/SOF3/include-flate/pull/25/commits/957c2bf5f9482651839b7db3f873175f51ba24cc))

kkent030315 commented:
I’ve also added compile-time warning for less compression binaries, so they have a visibility for compression ration for each binaries and can consider another compression method if it is not they're looking for.

Also for the users want to ignore this warning, I've added `no-compression-warnings` feature, which is disabled by default.

![image](https://github.com/SOF3/include-flate/assets/37926134/0b7b53b6-4ab9-46bd-ab67-049915a0d460)

SOF3 commented on codegen/src/lib.rs:
```diff
@@ -55,47 +59,117 @@ pub fn deflate_file(ts: TokenStream) -> TokenStream {
+impl syn::parse::Parse for FlateArgs {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let path = input.parse()?;
+        let algorithm = input.parse().ok();
```
Did you forget to parse `with` here?

kkent030315 replied:
Nope, `with` is not passed to this proc-macro.
This is just an user-friendly keyword so there is no rational reason to pass this in the proc-macro crate and implement redundant parsing code for it.
```rust
$crate::decode_string($crate::codegen::deflate_utf8_file!($path $($algo)?), Some($crate::CompressionMethodTy(algo)))
```

SOF3 commented on codegen/src/lib.rs:
```diff
@@ -55,47 +59,117 @@ pub fn deflate_file(ts: TokenStream) -> TokenStream {
+
+    fn try_from(value: Ident) -> Result<Self, Self::Error> {
+        match value.to_string().as_str() {
+            "deflate" => Ok(CompressionMethodTy(CompressionMethod::Deflate)),
```
What about just using `syn::custom_keyword` and `input.lookahead1()` for better diagnostics?

kkent030315 replied:
I do not think that is good idea. that would increase code dependency and brings more complexity.

kkent030315 replied:
Well, I've pushed the code that uses custom keyword as you wish. I also think there's no much pain for both solutions. 
05481fd

SOF3 commented on codegen/src/lib.rs:
```diff
@@ -55,47 +59,117 @@ pub fn deflate_file(ts: TokenStream) -> TokenStream {
+    }
+}
+
+fn calculate_compression_ratio(original_size: u64, compressed_size: u64) -> f64 {
```
```suggestion
fn compression_ratio(original_size: u64, compressed_size: u64) -> f64 {
```

kkent030315 commented:
@SOF3 For the compression ratio warning, we should have discussion about:
1. Should it be disabled by default? I think so.
2. What do you think the best threshold is? Currently, it is set to 10%.

SOF3 commented:
10% sounds right to me. Users should be aware that include_flate does not just increase executable size &mdash; it also increases the runtime memory by doubling the required storage and hence only reasonable to use when compression ratio is sufficiently high.

Might be useful to provide a deflate_if in a future (separate) feature too.

kkent030315 commented:
Okay, l will work on fixing CI.

kkent030315 commented:
It should be ok for now. please run the CI.

SOF3 commented on compress/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+description = "Compression algorithm provider"
+
+[dependencies]
+libflate = "2.0.0"
```
I think we still benefit from cfg features to avoid compiling unused algorithms though. Just don't use it for changing program behavior, but I think it's fine to use it for reducing unnecessary dependencies.

kkent030315 replied:
So you mean the user could *optionally* disable specific dependency for algorithms by their manual choise, if they don't use it, right?
Then making some features in both compression library crate and root crate like this would be an ideal option as previously implemented in early commits in this PR:

```toml
libflate = { version = "2.0.0", optional = true }
zstd = { version = "0.13.0", optional = true }

[features]
default = ["deflate", "zstd"]
deflate = ["dep:libflate"]
zstd = ["dep:zstd"]

# include-flate-compress/Cargo.toml
```

```toml
[features]
default = ["deflate", "zstd"]
deflate = ["include-flate-compress/deflate"]
zstd = ["include-flate-compress/zstd"]

# include-flate/Cargo.toml
```

And then, They're both set enabled by default, if user doesn't care about `deflate` method, like this:

```toml
[dependencies]
include-flate = { version = "...", default-features = false, features = ["zstd"] }
```

And vice versa:

```toml
[dependencies]
include-flate = { version = "...", default-features = false, features = ["deflate"] }
```

kkent030315 replied:
Done.

SOF3 commented:
Thanks! Will do a release later.
